emulator: Implement the DropRowRanges RPC.

marete opened the pull request:
TESTED=bigtable_mutations_integration_test now passes when this commit is merged to the integration branch.
    
References: TBL-59
Fixes: TBL-59

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9)
<!-- Reviewable:end -->

marete committed:
emulator: Start RowTransaction class.

Logic from Table::MutateRow() will be moved here and this commit
proposes the general structure of this class, for early review.

marete committed:
emulator: Implement undo log and wire-up rewind for the SetCell mutation.

- Introduces a RowTransaction class to proxy mutations and record undo
  logs in case we needs to rewind.
- Re-implements the SetCell mutation in the RowTransaction class.
- Introduces all Undo types/actions for SetCell and records them in an
  undo stack.
- For SetCell only and to facilitate testing with only SetCell
  mutations, introduces the use of RowTransaction in doing the mutation.
- clang-format is run on all affected files (which has removed some
  headers, e.t.c.

A test for SetCell is coming up and is not currently included here.

marete committed:
emulator: Properly declate all possible row mutations.

And unconditionally return an error status for the aggregation
mutations, which we don't currently support.

marete committed:
emulator: Implement a SetCell that records an undo log.

marete committed:
Complete implementation of Undo() for SetCell mutations.

marete committed:
emulator: Implement a constructor for the RowTransaction class.

marete committed:
emulator: Pass RowTransaction a constant reference to the mutation request.

marete committed:
emulator: Introduce the use of a RowTransaction in Table::MutateRow.

Currently in use only for SetCell.

marete committed:
emulator: Run clang-format on the subset of the files we have been changing.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator.

marete committed:
emulator: Start test suite for row mutation transaction rollback.

marete committed:
emulator: Add a static method to Table that returns a default-constructed table.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: testing: Create a table in the correct way.

Populate a schema protobuf with the minimal fields and call
Table::Create();

marete committed:
emulator: testing: rollback_test: Fix crash when setting up SetCell mutation.

There is a crash later on, in the code that sets up the undo log, that
will be fixed in an upcoming commit.

marete committed:
emulator: SetCell mutation: Various logic fixes for undo log.

marete committed:
emulator: transaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
emulator: trasnaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: table: Add a method to return an iterator to the column families map.

marete committed:
emulator: Introduce methods to retrieve begin and end of column family map iteration.

marete committed:
emulator: Complete implementing test for basic transaction functionality.

The test ensures that SetCell mutations still work via the
RowTransaction class.

Rollback will be tested in an upcoming change.

marete committed:
emulator: rollback test: Fix for more consistent handling of the table shared pointer.

marete committed:
emulator: Store a std::shared_ptr<Table> in RowTransaction.

This is a more correct way of doing it than storing a raw pointer.

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: ColumnFamily class: Disable copying.

marete committed:
emulator: Disable copying ColumnFamilyRow and simplify tx SetCell logic.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: Disable the copying of ColumnRows.

marete committed:
emulator: rollback_test.cc: factor out helper to check a cell exists with a specific value.

marete committed:
emulator: add helper to set a cell to a value.

Also, run clang-format on the file.

marete committed:
emulator: rollback tests: implement helper function to create tables.

marete committed:
emulator: rollack tests: Fix comment describing test.

marete committed:
emulator: rollback tests: Implement additional testing helper functions.

- To check if a row exists
- To check if a column exists

marete committed:
emulator: add helper to test > 1 SetCell mutations at once.

marete committed:
emulator: use set_cells() to set possibly more than one SetCell mutation.

marete committed:
emualtor: transaction rollback: Implement working test for RestoreValue.

This also involves a fix to the code due to a bug revealed by the
test!

marete committed:
emulator: rollback_tests.cc: run clang-format.

marete committed:
emulator: transaction rollback: Avoid copying the possibly very large cell values.

marete committed:
emulator: transaction rollback: SetCell: Implement working test for DeleteValue.

marete committed:
emulator: transaction rollback: Implement working DeleteRow test.

marete committed:
emulator: transaction rollback: Implement a DeleteFromRow that updates the undo log.

Also, add code coverage output artefacts to ignored files.

marete committed:
emulator: transaction rollback: Implement undo for row deletions (DeleteFromFamily).

And be careful not to potentially very large values (use std::move on
the column qualifier and cell values.

marete committed:
emulator: RestoreRow undo log: Rename struct members more conventionally.

marete committed:
emulator: DeleteFromFamily mutation: Implement with rollback support using RowTransaction.

marete committed:
emulator: transaction rollback: Test basic functionality of a DeleteFromFamily which uses RowTransaction.

marete committed:
emulator: test rollback of a row deletion due to DeleteFromFamily.

And fix a code bug discovered by this test.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes for review: Improve name of return varible.

In particular, make it clear that we are only examining the status,
and will not use the possibly valid status value (returned object).

marete committed:
emulator: Rename RestoreRow -> RestoreColumnFamilyRow.

marete committed:
emulator: style: Remove trailing underscore from struct member names.

marete committed:
emulator: Rename several iterator variables more appropriately.

marete committed:
emulator: Use std::move() to avoid a copy.

marete committed:
emulator: Rename another iterator properly.

marete committed:
emulator: Use std::move() to avoid another copy.

marete committed:
emulator: check and return from exceptional conditions first.

This is to remain in line with the rest of the code.

marete committed:
emualtor: Fix the naming of another iterator variable.

marete committed:
emulator: remove use of C++17 syntax feature (if with initializer).

We need to have the code compilable by a C++14 compiler.

marete committed:
emulator: rename another iterator variable more appropriately.

marete committed:
emulator: Add a comment justifying storing the column family iterator in the undo log.

In general, it is unsafe to store other kind of map iterators in the
undo log, since a subsequent mutation on the same row could delete the
underlying map, and invalidate the iterator. However, deleting a
column family is a schema operation that cannot happen while we hold
the table lock, therefore for column families only, it is safe to
store the iterators in the undo log.

marete committed:
emulator: transaction rollback: RestoreValue: Eliminate the storage of iterators in undo log.

marete committed:
emulator: actually remove iterator field from RestoreValue.

marete committed:
emulator: Correct and radically simplify SetCell undo code logic.

- Don't store iterators in the DeleteValue undo type (they may be
  invalidated by a subsequent mutation).

- Maintain the invariants that there is never an empty column or empty
  row by using the ColumnFamily::DeleteColumn() function to delete
  ranges of cells during undo. The latter takes care of deleting empty
  columns and empty rows, already.

- And therefore we can get rid of the DeleteRow and DeleteColumn undo
  types and associated code.

All tests continue to pass.

marete committed:
emulator: remove undo types that are no longer needed.

marete committed:
emulator: Delete definitions of rollback types no longer used.

marete committed:
emulator: return unimplemented error status for unsupported aggregation mutations.

Replaces FIXMEs which would silently fail when this mutations are attempted.

marete committed:
emulator: row mutations: use a const reference rather than copying the mutation.

marete committed:
emulator: Modify the Delete* methods to return vector of deleted cells.

And also use std::move() to avoid copying potentially large cell values.

These change is needed to simplify the logic of preparing undo logs
for the Delete* mutations. In particular, we don't have to re-iterate
the table.

marete committed:
emualator: Implement DeleteFromColumn mutation using RowTransaction.

marete committed:
emulator: Test transaction DeleteFromColumn for basic functionality.

marete committed:
emulator: DeleteFromColumn basic functionality test: test that the column is actually deleted.

marete committed:
emulator: make it possible to test a vector of DeleteFromColumn row mutations.

This is needed to test DeleteFromColumn rollback.

marete committed:
emulator: Test transaction rollback for DeleteFromColumn.

marete committed:
emulator: TEST(TransactonRollback, DeleteFromColumnRollback): also check that the cells are properly restored.

marete committed:
emulator: Prepare to support transaction rollbacks for DeleteRow mutation.

In particular, reutrn a map of vectors containing the column qualifers
to columns (cells) deleted, for the undo log.

marete committed:
emulator: DeleteRow: don't forget to erase the row key.

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

Also implement DeleteFromFamily in terms of the new DeleteRow which
returns RestoreValue vectors (and therefore eliminiate the use of one
more undo log type).

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: clang-format.

marete committed:
emulator: transaction rollback: get rid of another undo log type.

This leaves us with just 2 undo log types: RestoreValue and
DeleteValue, a situation that is much simpler than before.

marete committed:
emulator: Implement a DeleteFromRow that supports rollback.

marete committed:
emulator: Test DeleteFromRow with rollback support and use it in MutateRow code.

marete committed:
emulator: Update and correct a comment.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
erge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emualtor: Remove now-unused undo type.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes to avoid copying values that are possibly very large (cell contents).

As well as other fixes for review.

marete committed:
emulator: mutation transactions: use std::chrono:duration_cast.

Makes it easier and safer to change the resolution at a later date.

marete committed:
emulator: use invariant that rows are never empty to eliminate some code.

Preceding code erases all columns. The last column erase in
DeleteColumn will cause the row to be erased. So we don't have to
search for it again in the map and attempt to erase it.

marete committed:
emulator: optimization: make it possible to delete a column via a row iterator.

In this and other situations, this saves us a duplicate call to find
the row by saving and passing an iterator.

marete committed:
emulator: eliminate redundant storage of row_key in undo log.

The entire undo is in the context of a row, therefore all the methods
have access to a row key.

This also reduces the size of the undo log.

marete committed:
emulator: Use google-standard status builder everywhere in main code.

marete committed:
emulator: Add a FIXME to ask to use google-standard Status builder everywhere in test file.

marete committed:
emulator: Remove unused local var.

marete committed:
emulator: eliminate a C++-17ism.

marete committed:
emulator: DeleteRow: fix hang/crash due to use of invalidated row iterator.

marete committed:
emulator: remove non-Google-standard Status builders.

marete committed:
emulator: remove some outdated FIXMEs.

- We now have atomic rollback support for all implemented mutations.

marete committed:
emulator: make some std::chrono usage more type-safe.

marete committed:
emulator: std::move() to avoid another copy of a cell.

marete committed:
emulator: eliminate some code duplication.

marete committed:
emulator: SetCell: return overwritten value (if any).

This prepares the way to simplify the code that prepares the undo log,
and in particular it obviates the need to check to see if the
key (timestamp) existed before the SetCell mutation.

marete committed:
emulator: radically simplify undo log preparation code.

This is thanks to simply returning the old value (if any) rather than
performing almost endless iterations in this function! :)

marete committed:
emulator: ColumnFamily::DeleteRow(): radically simplify the function.

And also eliminate the risk of accessing an invalidated iterator.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: SetCell: set timestamp to server time if passed 0 or negative timestamp.

marete committed:
emulator: transaction rollback: Check that a SetCell with timestamp 0 can be correctly rolled back.

marete committed:
emulator: docs: correct formatting of several comments in header.

marete committed:
emulator: tests: use a better EXPECT* macro in one place.

marete committed:
emulator: Implement ColumnFamily::DeleteTimeStamp().

This allows us to delete cells by specifying a specific timestamp
rather than a timestamp range which might otherwise involve possibly
non-type-safe calculations to determine the upper bound of the
timestamps (in particular doing the latter needs one to assume a
specific resolution for bigtable's timestamp storage).

marete committed:
emulator: transaction undo: radically simplify the code for DeleteValue.

marete committed:
emulator: rollback tests: perform some trivial refactoring.

marete committed:
emulator tests: Fix bigtable_emulator_range_set_test.

The tests aborted due to an assertion in the constructor of Range that
enforced the invariant that start <= end for all well-formed
Ranges. However in several tests testing the ends of ranges, start was
passed in as "unimportant" with end being "A" or "B". This violates
the invariant and causes an abort at runtime.

Fixes: TBL-62.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: fixes for clang-tidy and clang-format.

Restricted only to changes I have made to this PR.

Tested (all tests pass).

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' into fix-test-bigtable_emulator_range_set_test

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Implement the DropRowRanges RPC.

TESTED=bigtable_mutations_integration_test now passes when this commit
is merged to the integration branch.

References: TBL-59
Fixes: TBL-59

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9)* status: 0 of 4 files reviewed, 2 unresolved discussions

___
*[`google/cloud/bigtable/emulator/table.cc` line 426 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQ8tNZS3U6KzDbIdSKD:-OQ8tNZS3U6KzDbIdSKE:b-b7hszu) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/efca9cc41dbdbb60d8c4a6375f634852097cf72a/google/cloud/bigtable/emulator/table.cc#L426)):*
> ```cpp
> 
>   for (auto& cf : column_families_) {
>     for (auto row_it = cf.second->begin_mutable(); row_it != cf.second->end();
> ```

We don't need to check all rows to remove them by prefix. They are sorted, so starting at `std::map::lower_bound(prefix)` and finishing as soon as you find a row which does not match the prefix sounds much more efficient.
___
*[`google/cloud/bigtable/emulator/table.cc` line 429 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQ8sgGB7HCBB9JzIPgi:-OQ8sgGB7HCBB9JzIPgj:b-m6164j) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/efca9cc41dbdbb60d8c4a6375f634852097cf72a/google/cloud/bigtable/emulator/table.cc#L429)):*
> ```cpp
>          row_it++) {
>       if (absl::StartsWith(row_it->first, row_prefix)) {
>         cf.second->erase(row_it);
> ```

This will invalidate the `row_it` iterator so the for loop will be advancing an invalidated iterator.



_Code quote:_
```cpp
        cf.second->erase(row_it
```


<!-- Sent from Reviewable.io -->

dopiera reviewed (commented):
Also, can you please add a test to make sure that we ara really erasing the right rows?

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-:-OQ8uiu_CuYyC1m3yld1:b4pe5z1)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @marete)



<!-- Sent from Reviewable.io -->

marete committed:
Merge branch 'upstream_emulator' into drop_row_range

marete committed:
emulator: DropRowRange: Avoid iterating over the entire table to find matches.

And also, fix an almost certain crash resulting from reusing an
erased (invalidated) iterator.

marete committed:
emulator: DropRowRange: Test option to drop all.

marete committed:
emulator: DropRowRange: mechanical fixes by CI tool.

marete committed:
emulator: DropRowRange: Fix the tests and fix a bug uncovered by the tests.

marete reviewed (commented):
Thanks for insisting on this. The tests discovered that the code did not work when a range was given. Fixed Now.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-:-OQARUnZ4A5FzolDimuH:b-cih782)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/table.cc` line 426 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQ8tNZS3U6KzDbIdSKD:-OQ9cFBJ9MhL0IjXCNxo:bgfmch7) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/efca9cc41dbdbb60d8c4a6375f634852097cf72a/google/cloud/bigtable/emulator/table.cc#L426)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

We don't need to check all rows to remove them by prefix. They are sorted, so starting at `std::map::lower_bound(prefix)` and finishing as soon as you find a row which does not match the prefix sounds much more efficient.
</blockquote></details>

Many thanks for pointing out this important optimization. Fixed.
___
*[`google/cloud/bigtable/emulator/table.cc` line 429 at r1](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQ8sgGB7HCBB9JzIPgi:-OQ9cLOn4qhRTDS75MLX:bm0yrq6) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/efca9cc41dbdbb60d8c4a6375f634852097cf72a/google/cloud/bigtable/emulator/table.cc#L429)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

This will invalidate the `row_it` iterator so the for loop will be advancing an invalidated iterator.
</blockquote></details>

Thanks for noticing this. Now using the C++11 feature in which erase returns the iterator to the next element in the container.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
PTAL. Thanks!

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-:-OQARvli9TWvN_ufo_Ax:b-lxcnv2)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9)* status: 0 of 7 files reviewed, 5 unresolved discussions

___
*[`google/cloud/bigtable/emulator/column_family.h` line 204 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCbKsF0hdoZKsKKt3T:-OQCbKsF0hdoZKsKKt3U:bntdv0z) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/column_family.h#L204)):*
> ```objc
> 
>   using const_iterator = std::map<std::string, ColumnFamilyRow>::const_iterator;
>   using mutable_iterator = std::map<std::string, ColumnFamilyRow>::iterator;
> ```

Sorry - I should have pointed it out earlier. Let's stick to the standard naming scheme, which is `iterator` and `const_iterator`, i.e. please drop `mutable`.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 275 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCbZ1Z5Z3cZ8_fqmcc:-OQCbZ1Z5Z3cZ8_fqmcd:btn0w62) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/column_family.h#L275)):*
> ```objc
> 
>   const_iterator begin() const { return rows_.begin(); }
>   mutable_iterator begin_mutable() { return rows_.begin(); }
> ```

ditto for these - please also call then `begin()` - the compiler should pick the right overload depending on whether you've got a const reference or not.



_Code quote:_
```objc
 mutable_iterator begin_mutable() 
```
___
*[`google/cloud/bigtable/emulator/drop_row_range_test.cc` line 50 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCc8wv2tpgiz-b41zR:-OQCc8wv2tpgiz-b41zS:btqla2o) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/drop_row_range_test.cc#L50)):*
> ```cpp
> };
> 
> StatusOr<std::shared_ptr<Table>> create_table(
> ```

Sorry I didn't notice it earlier - the function names don't follow the guideline (CamelCase) - can you please make sure they do?


<!-- Sent from Reviewable.io -->

marete committed:
emulator: DropRowRange: Fix naming of test functions - use PascalCase.

marete committed:
emulator: ColumnFamily: Use const-correctness to provide alternative mutator iterators.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/column_family.h` line 204 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCbKsF0hdoZKsKKt3T:-OQDeJt61Cmulu9obksz:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/column_family.h#L204)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Sorry - I should have pointed it out earlier. Let's stick to the standard naming scheme, which is `iterator` and `const_iterator`, i.e. please drop `mutable`.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 275 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCbZ1Z5Z3cZ8_fqmcc:-OQDeKdY5fYfT3UlTiTG:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/column_family.h#L275)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

ditto for these - please also call then `begin()` - the compiler should pick the right overload depending on whether you've got a const reference or not.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/drop_row_range_test.cc` line 50 at r2](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-OQCc8wv2tpgiz-b41zR:-OQCzjHoDiWxLFHGIqwn:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6ee107aa8da877b77af4d693bf3ca6af1b85669f/google/cloud/bigtable/emulator/drop_row_range_test.cc#L50)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Sorry I didn't notice it earlier - the function names don't follow the guideline (CamelCase) - can you please make sure they do?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
Done, PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/9#-:-OQDeQmsD0EXD9pZjeip:bevftus)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->
